File: app/src/main/java/com/owncloud/android/ui/adapter/ActivityAndVersionListAdapter.java
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2020 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2019 Chris Narkiewicz <hello@ezaquarii.com>
 * SPDX-FileCopyrightText: 2019 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.ui.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.nextcloud.client.account.CurrentAccountProvider;
import com.nextcloud.client.network.ClientFactory;
import com.nextcloud.common.NextcloudClient;
import com.owncloud.android.databinding.VersionListItemBinding;
import com.owncloud.android.lib.resources.activities.model.Activity;
import com.owncloud.android.lib.resources.files.model.FileVersion;
import com.owncloud.android.ui.interfaces.ActivityListInterface;
import com.owncloud.android.ui.interfaces.VersionListInterface;
import com.owncloud.android.utils.DisplayUtils;
import com.owncloud.android.utils.theme.ViewThemeUtils;

import java.util.Collections;
import java.util.Date;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class ActivityAndVersionListAdapter extends ActivityListAdapter {
    private static final int VERSION_TYPE = 102;

    private final VersionListInterface.View versionListInterface;

    public ActivityAndVersionListAdapter(
        Context context,
        CurrentAccountProvider currentAccountProvider,
        ActivityListInterface activityListInterface,
        VersionListInterface.View versionListInterface,
        ClientFactory clientFactory,
        ViewThemeUtils viewThemeUtils
                                        ) {
        super(context,
              currentAccountProvider,
              activityListInterface,
              clientFactory,
              true,
              viewThemeUtils);

        this.versionListInterface = versionListInterface;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setActivityAndVersionItems(List<Object> items, NextcloudClient newClient, boolean clear) {
        if (client == null) {
            client = newClient;
        }
        if (clear) {
            values.clear();
            items.sort((o1, o2) -> {
                long o1Date;
                long o2Date;
                if (o1 instanceof Activity) {
                    o1Date = ((Activity) o1).getDatetime().getTime();
                } else {
                    o1Date = ((FileVersion) o1).getModifiedTimestamp();
                }

                if (o2 instanceof Activity) {
                    o2Date = ((Activity) o2).getDatetime().getTime();
                } else {
                    o2Date = ((FileVersion) o2).getModifiedTimestamp();
                }

                return -1 * Long.compare(o1Date, o2Date);
            });
        }

        String sTime = "";
        for (Object item : items) {
            String time;

            if (item instanceof Activity activity) {
                time = getHeaderDateString(context, activity.getDatetime().getTime()).toString();
            } else {
                FileVersion version = (FileVersion) item;
                time = getHeaderDateString(context, version.getModifiedTimestamp()).toString();
            }

            if (sTime.equalsIgnoreCase(time)) {
                values.add(item);
            } else {
                sTime = time;
                values.add(sTime);
                values.add(item);
            }
        }

        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VERSION_TYPE) {
            return new VersionViewHolder(VersionListItemBinding.inflate(LayoutInflater.from(parent.getContext())));
        }

        return super.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof VersionViewHolder versionViewHolder) {
            FileVersion fileVersion = (FileVersion) values.get(position);

            versionViewHolder.binding.size.setText(DisplayUtils.bytesToHumanReadable(fileVersion.getFileLength()));
            versionViewHolder.binding.time.setText(
                DateFormat.format("HH:mm", new Date(fileVersion.getModifiedTimestamp()).getTime())
            );

            versionViewHolder.binding.restore.setOnClickListener(
                v -> versionListInterface.onRestoreClicked(fileVersion)
            );
        } else {
            super.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemViewType(int position) {
        Object value = values.get(position);

        if (value instanceof Activity) {
            return ACTIVITY_TYPE;
        } else if (value instanceof FileVersion) {
            return VERSION_TYPE;
        } else {
            return HEADER_TYPE;
        }
    }

    protected class VersionViewHolder extends RecyclerView.ViewHolder {

        VersionListItemBinding binding;

        VersionViewHolder(VersionListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/owncloud/android/ui/activities/adapter/ActivityAndVersionListAdapter.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
package com.owncloud.android.ui.activities.adapter

import android.annotation.SuppressLint
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.nextcloud.client.account.CurrentAccountProvider
import com.nextcloud.common.NextcloudClient
import com.owncloud.android.databinding.VersionListItemBinding
import com.owncloud.android.lib.resources.activities.model.Activity
import com.owncloud.android.lib.resources.files.model.FileVersion
import com.owncloud.android.ui.interfaces.ActivityListInterface
import com.owncloud.android.ui.interfaces.VersionListInterface
import com.owncloud.android.utils.DisplayUtils
import com.owncloud.android.utils.theme.ViewThemeUtils
import java.util.Date

class ActivityAndVersionListAdapter(
    context: FragmentActivity,
    currentAccountProvider: CurrentAccountProvider,
    activityListInterface: ActivityListInterface,
    private val versionListInterface: VersionListInterface.View,
    viewThemeUtils: ViewThemeUtils
) : ActivityListAdapter(context, currentAccountProvider, activityListInterface, true, viewThemeUtils) {

    @SuppressLint("NotifyDataSetChanged")
    fun setActivityAndVersionItems(items: MutableList<Any?>, newClient: NextcloudClient?, clear: Boolean) {
        if (client == null) client = newClient

        if (clear) {
            values.clear()
            items.sortByDescending { it.timestamp() }
        }

        var sTime = ""
        for (item in items) {
            val time = getHeaderDateString(context, item.timestamp() ?: continue).toString()
            if (!sTime.equals(time, ignoreCase = true)) {
                sTime = time
                values.add(sTime)
            }
            if (item != null) {
                values.add(item)
            }
        }

        notifyDataSetChanged()
    }

    private fun Any?.timestamp(): Long? = when (this) {
        is Activity -> datetime.time
        is FileVersion -> modifiedTimestamp
        else -> null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VERSION_TYPE) {
            return VersionViewHolder(VersionListItemBinding.inflate(LayoutInflater.from(parent.context)))
        }
        return super.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is VersionViewHolder) {
            val fileVersion = values[position] as FileVersion
            holder.binding.size.text = DisplayUtils.bytesToHumanReadable(fileVersion.fileLength)
            holder.binding.time.text = DateFormat.format("HH:mm", Date(fileVersion.modifiedTimestamp).time)
            holder.binding.restore.setOnClickListener { versionListInterface.onRestoreClicked(fileVersion) }
        } else {
            super.onBindViewHolder(holder, position)
        }
    }

    override fun getItemViewType(position: Int) = when (values[position]) {
        is Activity -> ACTIVITY_TYPE
        is FileVersion -> VERSION_TYPE
        else -> HEADER_TYPE
    }

    open class VersionViewHolder(val binding: VersionListItemBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        private const val VERSION_TYPE = 102
    }
}
